File: lint-arwaky/src-python/agent/pipeline_action_orchestrator.py
"""
pipeline_action_orchestrator — Logic for dispatching pipeline actions (Agent Layer).
"""

from __future__ import annotations
from typing import Mapping, Callable
from ..contract import OrchestratorContainerAggregate, PipelineActionDispatcherAggregate
from ..taxonomy import (
    ContentString,
    MetadataVO,
    FilePath,
    BooleanVO,
    ResponseData,
    VALID_PIPELINE_ACTIONS,
)


class PipelineActionDispatcher(PipelineActionDispatcherAggregate):
    """Dispatches actions to the appropriate orchestrators or capabilities."""

    def __init__(self, container: OrchestratorContainerAggregate, **data: object) -> None:
        super().__init__(container=container, **data)

    async def dispatch(self, action: ContentString, args: MetadataVO) -> ResponseData:
        """Dispatch action to the appropriate use case or tool via handler map."""
        action_str = str(action.value) if hasattr(action, "value") else str(action)

        handler_map: Mapping[str, Callable[[str, MetadataVO], ResponseData]] = {
            "check": self._handle_check,
            "scan": self._handle_check,
            "security": self._handle_security,
            "complexity": self._handle_complexity,
            "duplicates": self._handle_duplicates,
            "trends": self._handle_trends,
            "fix": self._handle_fix,
            "report": self._handle_report,
            "version": self._handle_version,
            "adapters": self._handle_adapters,
            "install-hook": self._handle_install_hook,
            "install_hook": self._handle_install_hook,
            "uninstall-hook": self._handle_uninstall_hook,
            "uninstall_hook": self._handle_uninstall_hook,
            "batch": self._handle_batch,
            "multi_project": self._handle_multi_project,
            "doctor": self._handle_doctor,
            "cancel": self._handle_cancel,
        }

        handler = handler_map.get(action_str)
        if handler is not None:
            return await handler(action_str, args)

        return ResponseData(
            data={"error": f"No pipeline handler for action: {action_str}"}
        )

    # ------------------------------------------------------------------
    # Handler methods (one per action)
    # ------------------------------------------------------------------

    @staticmethod
    def _extract_path(args: MetadataVO) -> FilePath:
        """Extract FilePath from MetadataVO args."""
        path_val = (
            args.data.get("path", ".")
            if hasattr(args, "data")
            else args.get("path", ".")
        )
        return FilePath(value=str(path_val))

    async def _handle_check(self, action_str: str, args: MetadataVO) -> ResponseData:
        path = self._extract_path(args)
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        return ResponseData(data=data)

    async def _handle_security(self, action_str: str, args: MetadataVO) -> ResponseData:
        path = self._extract_path(args)
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        return ResponseData(data={"bandit": data.get("bandit", [])})

    async def _handle_complexity(
        self, action_str: str, args: MetadataVO
    ) -> ResponseData:
        path = self._extract_path(args)
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        return ResponseData(data={"radon": data.get("radon", [])})

    async def _handle_duplicates(
        self, action_str: str, args: MetadataVO
    ) -> ResponseData:
        path = self._extract_path(args)
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        return ResponseData(data={"duplicates": data.get("duplicates", [])})

    async def _handle_trends(self, action_str: str, args: MetadataVO) -> ResponseData:
        path = self._extract_path(args)
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        return ResponseData(data={"trends": data.get("trends", [])})

    async def _handle_fix(self, action_str: str, args: MetadataVO) -> ResponseData:
        path = self._extract_path(args)
        output = await self.container.fix_orchestrator.execute(path)
        return ResponseData(data={"output": str(output)})

    async def _handle_report(self, action_str: str, args: MetadataVO) -> ResponseData:
        path = self._extract_path(args)
        output_format = (
            str(args.data.get("format", "text"))
            if hasattr(args, "data")
            else str(args.get("format", "text"))
        )
        report = await self.container.analysis_orchestrator.run(path)
        data = self.container.to_dict(report)
        formatters = {
            "json": lambda r, d: {"format": "json", "data": d},
            "sarif": lambda r, d: {
                "format": "sarif",
                "data": self.container.to_sarif_report(r).value,
            },
            "junit": lambda r, d: {
                "format": "junit",
                "data": self.container.to_junit_report(r).value,
            },
        }
        formatter = formatters.get(
            output_format, lambda r, d: {"format": "text", "data": d}
        )
        return ResponseData(data=formatter(report, data))

    async def _handle_version(self, action_str: str, args: MetadataVO) -> ResponseData:
        return ResponseData(data=await self._get_version())

    async def _handle_adapters(self, action_str: str, args: MetadataVO) -> ResponseData:
        return ResponseData(
            data={"adapters": [str(a.name()) for a in self.container.adapters]}
        )

    async def _handle_install_hook(
        self, action_str: str, args: MetadataVO
    ) -> ResponseData:
        ok = self.container.hook_capability.install()
        return ResponseData(data={"installed": ok})

    async def _handle_uninstall_hook(
        self, action_str: str, args: MetadataVO
    ) -> ResponseData:
        ok = self.container.hook_capability.uninstall()
        return ResponseData(data={"uninstalled": ok})

    # Future / placeholder handlers
    async def _handle_batch(self, action_str: str, args: MetadataVO) -> ResponseData:
        return ResponseData(data={"error": "batch action not yet implemented"})

    async def _handle_multi_project(
        self, action_str: str, args: MetadataVO
    ) -> ResponseData:
        return ResponseData(data={"error": "multi_project action not yet implemented"})

    async def _handle_doctor(self, action_str: str, args: MetadataVO) -> ResponseData:
        return ResponseData(data={"error": "doctor action not yet implemented"})

    async def _handle_cancel(self, action_str: str, args: MetadataVO) -> ResponseData:
        return ResponseData(data={"error": "cancel action not yet implemented"})

    async def _get_version(self) -> Mapping[str, str]:
        from importlib.metadata import version as _v, PackageNotFoundError

        try:
            ver = _v("auto-linter")
        except PackageNotFoundError:
            ver = "1.0.0"
        return {"version": ver}

    def validate_action(self, action: ContentString) -> BooleanVO:
        """Check if action is known — uses canonical VALID_PIPELINE_ACTIONS."""
        return BooleanVO(
            value=str(action.value if hasattr(action, "value") else action)
            in VALID_PIPELINE_ACTIONS
        )

File: lint-arwaky/src-rust/agent/pipeline_action_orchestrator.rs
// pipeline_action_orchestrator — Logic for dispatching pipeline actions (Agent Layer).
use crate::contract::{PipelineDispatcherAggregate};
use crate::taxonomy::{ContentString, MetadataVO, FilePath, BooleanVO, ResponseData, StdOutput, StdError, ExitCode};
use std::collections::HashMap;

pub struct PipelineActionDispatcher;

impl PipelineDispatcherAggregate for PipelineActionDispatcher {}

impl PipelineActionDispatcher {
    pub fn new() -> Self {
        Self
    }

    pub async fn dispatch(&self, action: &ContentString, _args: &MetadataVO) -> ResponseData {
        let action_str = &action.value;

        // Handler map for known actions
        match action_str.as_str() {
            "check" | "scan" => self.handle_check(action_str).await,
            "security" => self.handle_security(action_str).await,
            "complexity" => self.handle_complexity(action_str).await,
            "duplicates" => self.handle_duplicates(action_str).await,
            "trends" => self.handle_trends(action_str).await,
            "fix" => self.handle_fix(action_str).await,
            "report" => self.handle_report(action_str).await,
            "version" => self.handle_version(action_str).await,
            "adapters" => self.handle_adapters(action_str).await,
            "install-hook" | "install_hook" => self.handle_install_hook(action_str).await,
            "uninstall-hook" | "uninstall_hook" => self.handle_uninstall_hook(action_str).await,
            _ => {
                let mut metadata = HashMap::new();
                metadata.insert("error".to_string(), serde_json::json!(format!("No pipeline handler for action: {}", action_str)));
                ResponseData::new(
                    StdOutput::new(""),
                    StdError::new(format!("No pipeline handler for action: {}", action_str)),
                    ExitCode::new(1),
                    MetadataVO::new(metadata),
                )
            }
        }
    }

    async fn handle_check(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("message".to_string(), serde_json::json!("check completed"));
        ResponseData::new(StdOutput::new("check completed"), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_security(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("bandit".to_string(), serde_json::json!(Vec::<String>::new()));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_complexity(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("radon".to_string(), serde_json::json!(Vec::<String>::new()));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_duplicates(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("duplicates".to_string(), serde_json::json!(Vec::<String>::new()));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_trends(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("trends".to_string(), serde_json::json!(Vec::<String>::new()));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_fix(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("output".to_string(), serde_json::json!("fix applied"));
        ResponseData::new(StdOutput::new("fix applied"), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_report(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("format".to_string(), serde_json::json!("text"));
        metadata.insert("data".to_string(), serde_json::json!({}));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_version(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("version".to_string(), serde_json::json!("1.0.0"));
        ResponseData::new(StdOutput::new("1.0.0"), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_adapters(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("adapters".to_string(), serde_json::json!(Vec::<String>::new()));
        ResponseData::new(StdOutput::new(""), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_install_hook(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("installed".to_string(), serde_json::json!(true));
        ResponseData::new(StdOutput::new("hook installed"), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    async fn handle_uninstall_hook(&self, _action: &str) -> ResponseData {
        let mut metadata = HashMap::new();
        metadata.insert("uninstalled".to_string(), serde_json::json!(true));
        ResponseData::new(StdOutput::new("hook uninstalled"), StdError::new(""), ExitCode::new(0), MetadataVO::new(metadata))
    }

    pub fn validate_action(&self, action: &ContentString) -> BooleanVO {
        let known_actions = [
            "check", "scan", "security", "complexity", "duplicates", "trends",
            "fix", "report", "version", "adapters", "install-hook", "install_hook",
            "uninstall-hook", "uninstall_hook", "batch", "multi_project", "doctor", "cancel",
        ];
        BooleanVO::new(known_actions.contains(&action.value.as_str()))
    }
}
